src/ray/gcs/asio.cc

```diff
@@ -0,0 +1,110 @@
+#include "asio.h"
+
+RedisAsioClient::RedisAsioClient(boost::asio::io_service &io_service,
+                                 redisAsyncContext *ac)
+    : context_(ac),
+      socket_(io_service),
+      read_requested_(false),
+      write_requested_(false),
+      read_in_progress_(false),
+      write_in_progress_(false) {
+  // gives access to c->fd
+  redisContext *c = &(ac->c);
+
+  // hiredis is already connected
+  // use the existing native socket
+  socket_.assign(boost::asio::ip::tcp::v4(), c->fd);
+
+  // register hooks with the hiredis async context
+  ac->ev.addRead = call_C_addRead;
+  ac->ev.delRead = call_C_delRead;
+  ac->ev.addWrite = call_C_addWrite;
+  ac->ev.delWrite = call_C_delWrite;
+  ac->ev.cleanup = call_C_cleanup;
+
+  // C wrapper functions will use this pointer to call class members.
+  ac->ev.data = this;
+}
+
+void RedisAsioClient::operate() {
+  if (read_requested_ && !read_in_progress_) {
+    read_in_progress_ = true;
+    socket_.async_read_some(boost::asio::null_buffers(),
+                            boost::bind(&RedisAsioClient::handle_read, this,
+                                        boost::asio::placeholders::error));
+  }
+
+  if (write_requested_ && !write_in_progress_) {
+    write_in_progress_ = true;
+    socket_.async_write_some(boost::asio::null_buffers(),
+                             boost::bind(&RedisAsioClient::handle_write, this,
+                                         boost::asio::placeholders::error));
+  }
+}
+
+void RedisAsioClient::handle_read(boost::system::error_code ec) {
+  read_in_progress_ = false;
+  if (!ec) {
+    redisAsyncHandleRead(context_);
+  }
+
+  if (!ec || ec == boost::asio::error::would_block) {
+    operate();
+  }
+}
```

Comment:
If we reach this part of the code that means that an error occurred, right? So maybe we should have a check failure if we get here (similarly with the other methods).

The downside is potential verbosity at exit, but maybe we can have the redis clients disconnect before killing the redis server to get around this.